src/server/worker/datum/create_stream.go

```diff
@@ -78,13 +95,466 @@ func streamingCreatePFS(ctx context.Context, c pfs.APIClient, taskDoer task.Doer
 			}
 			select {
 			case <-ctx.Done():
-				return errors.Wrap(context.Cause(ctx), "send file set id to iterator")
+				return errors.Wrap(context.Cause(ctx), "send file set id up")
 			case fsidChan <- result.FileSetId:
 			}
+		}
+		return nil
+	}); err != nil {
+		select {
+		case <-ctx.Done():
+		case errChan <- errors.Wrap(err, "streamingCreatePFS"):
+		}
+	}
+}
+
+func streamingCreateUnion(
+	ctx context.Context,
+	c pfs.APIClient,
+	taskDoer task.Doer,
+	inputs []*pps.Input,
+	fsidChan chan<- string,
+	errChan chan<- error,
+	requestDatumsChan <-chan struct{},
+) {
+	defer close(fsidChan)
+	if err := client.WithRenewer(ctx, c, func(ctx context.Context, renewer *renew.StringSet) error {
+		childrenFsidChans := initChildrenFsidChans(len(inputs))
+		go streamingCreateInputs(ctx, c, taskDoer, renewer, inputs, childrenFsidChans, errChan, requestDatumsChan)
+		if err := consumeChildrenFsidChans(ctx, childrenFsidChans, func(_ int, fsid string) error {
+			fsidChan <- fsid
+			return nil
+		}); err != nil {
+			return errors.Wrap(err, "consuming children channels")
+		}
+		return nil
+	}); err != nil {
+		select {
+		case <-ctx.Done():
+		case errChan <- errors.Wrap(err, "streamingCreateUnion"):
+		}
+	}
+}
+
+// Creates a slice of channels for each child input. Each channel will receive file set id shards
+func initChildrenFsidChans(num int) []chan string {
+	childrenFsidChans := make([]chan string, num)
+	for i := range num {
+		childrenFsidChans[i] = make(chan string)
+	}
+	return childrenFsidChans
+}
+
+// Consumes file set id shards from each child input as they arrive. Calls cb with the index of
+// the child input. Function returns when all children channels are closed or the context is done.
+func consumeChildrenFsidChans(ctx context.Context, childrenFsidChans []chan string, cb func(int, string) error) error {
+	cases := make([]reflect.SelectCase, len(childrenFsidChans)+1)
+	for i, childChan := range childrenFsidChans {
+		cases[i] = reflect.SelectCase{Dir: reflect.SelectRecv, Chan: reflect.ValueOf(childChan)}
+	}
+	cases[len(cases)-1] = reflect.SelectCase{Dir: reflect.SelectRecv, Chan: reflect.ValueOf(ctx.Done())}
+	finished := make(map[int]bool)
+	for len(finished) < len(cases)-1 {
+		i, value, ok := reflect.Select(cases)
+		if i == len(cases)-1 {
+			return errors.Wrap(context.Cause(ctx), "consumeChildrenFsidChans")
+		}
+		if ok {
+			fsid := value.Interface().(string)
+			if err := cb(i, fsid); err != nil {
+				return errors.Wrap(err, "consumeChildrenFsidChans")
+			}
+		} else {
+			finished[i] = true
```

Comment:
I think you need to remove the i-th selector from cases; nothing guarantees not hitting this 'matched because the channel is closed' case on the next Select call.  It could happen forever.

Reply:
Is the concern that we'll never exit from the loop or that we'll unnecessarily do extra iterations by matching on the closed channel repeatedly?